From c16846bf9e0b90521029f0d743620f138ee2236e Mon Sep 17 00:00:00 2001
From: Eric Hodel <drbrain@segment7.net>
Date: Wed, 9 Jul 2014 16:48:48 -0700
Subject: [PATCH] Generate docs when ignoring dependencies

During the switch to the new resolver documentation generation was
bypassed when the --ignore-dependencies option was given.

Now the done_installing hook is triggered when the
--ignore-dependencies option is present.

Fixes #961
---
 lib/rubygems/commands/install_command.rb           | 8 ++++++++
 test/rubygems/test_gem_commands_install_command.rb | 7 +++++++
 3 files changed, 17 insertions(+)

diff --git rubygems/lib/rubygems/commands/install_command.rb rubygems/lib/rubygems/commands/install_command.rb
index bc12aa1..1bf5928 100644
--- ruby/lib/rubygems/commands/install_command.rb
+++ ruby/lib/rubygems/commands/install_command.rb
@@ -278,6 +278,14 @@ def install_gem_without_dependencies name, req # :nodoc:
     inst = Gem::Installer.new gem, options
     inst.install
 
+    require 'rubygems/dependency_installer'
+    dinst = Gem::DependencyInstaller.new options
+    dinst.installed_gems.replace [inst.spec]
+
+    Gem.done_installing_hooks.each do |hook|
+      hook.call dinst, [inst.spec]
+    end unless Gem.done_installing_hooks.empty?
+
     @installed_specs.push(inst.spec)
   end
 
diff --git rubygems/test/rubygems/test_gem_commands_install_command.rb rubygems/test/rubygems/test_gem_commands_install_command.rb
index 857ba36..f03285a 100644
--- ruby/test/rubygems/test_gem_commands_install_command.rb
+++ ruby/test/rubygems/test_gem_commands_install_command.rb
@@ -565,6 +565,11 @@ def test_execute_conservative
   end
 
   def test_install_gem_ignore_dependencies_both
+    done_installing = false
+    Gem.done_installing do
+      done_installing = true
+    end
+
     spec = quick_spec 'a', 2
 
     util_build_gem spec
@@ -576,6 +581,8 @@ def test_install_gem_ignore_dependencies_both
     @cmd.install_gem 'a', '>= 0'
 
     assert_equal %w[a-2], @cmd.installed_specs.map { |s| s.full_name }
+
+    assert done_installing, 'documentation was not generated'
   end
 
   def test_install_gem_ignore_dependencies_remote
-- 
1.9.3

